### PR TITLE
Widget Top Interface - Option to ignore lo Interface

### PR DESCRIPTION
**Pull request by SourceDoctor**
Option to ignore Loopback Interfaces on Top Interfaces

image

#### Please note

> Please read this information carefully. You can run `./scripts/pre-commit.php` to check your code before submitting.

- [x] Have you followed our [code guidelines?](http://docs.librenms.org/Developing/Code-Guidelines/)
- [x] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by murrant** [comment]
Can we just have it respect disabled ports?

**Comment by SourceDoctor** [comment]
you mean to have the option to ignore disabled ports also?

**Comment by SourceDoctor** [comment]
@murrant 
hope this is what you meant

File: app/Http/Controllers/Widgets/TopInterfacesController.php [hunk 1]
```diff
@@ -54,11 +55,15 @@ public function getView(Request $request)
             ->select('port_id', 'device_id', 'ifName', 'ifDescr', 'ifAlias')
             ->groupBy('port_id', 'device_id', 'ifName', 'ifDescr', 'ifAlias')
             ->where('poll_time', '>', Carbon::now()->subMinutes($data['time_interval'])->timestamp)
+            ->where([['ignore', '=', 0], ['disabled', '=', 0]])
```
Comment (Jellyfrog) [inline app/Http/Controllers/Widgets/TopInterfacesController.php]: Use the scope. https://github.com/librenms/librenms/blob/master/app/Models/Port.php#L163

File: app/Http/Controllers/Widgets/TopInterfacesController.php [hunk 2]
```diff
@@ -54,11 +55,15 @@ public function getView(Request $request)
...
             ->when($data['device_group'], function ($query) use ($data) {
                 $query->inDeviceGroup($data['device_group']);
             }, function ($query) {
                 $query->has('device');
             })
+            ->when(!empty($data['hide_lo']), function ($query) {
+                $query->where('ifType', '<>', 'softwareLoopback');
```
Comment (Jellyfrog) [inline app/Http/Controllers/Widgets/TopInterfacesController.php]: This should probably be made into a scope. Sounds usable

File: app/Http/Controllers/Widgets/TopInterfacesController.php (hunk 1 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Widgets/TopInterfacesController.php]: how do i use it?

Comment (Jellyfrog) [reply app/Http/Controllers/Widgets/TopInterfacesController.php]: https://laravel.com/docs/7.x/eloquent#query-scopes

File: app/Http/Controllers/Widgets/TopInterfacesController.php (hunk 2 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Widgets/TopInterfacesController.php]: changed :-)

File: app/Http/Controllers/Widgets/TopInterfacesController.php (hunk 1 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Widgets/TopInterfacesController.php]: changed :)

File: app/Http/Controllers/Widgets/TopInterfacesController.php [hunk 3]
```diff
@@ -54,11 +55,15 @@ public function getView(Request $request)
             ->select('port_id', 'device_id', 'ifName', 'ifDescr', 'ifAlias')
             ->groupBy('port_id', 'device_id', 'ifName', 'ifDescr', 'ifAlias')
             ->where('poll_time', '>', Carbon::now()->subMinutes($data['time_interval'])->timestamp)
+            ->IsActive()
```
Comment (murrant) [inline app/Http/Controllers/Widgets/TopInterfacesController.php]: FYI, the case should be `->isActive()` for scopes.

**Comment by murrant** [comment]
@SourceDoctor if we respect disabled ports, doesn't that make the loopback option unnecessary?

**Comment by SourceDoctor** [comment]
if you disable a loopback Port on your system, your system will crash, so it cannot be disabled
so a loopback port is active

**Comment by murrant** [comment]
That isn't what disabled means.
ifOperStatus shows port up/down
ifAdminStatus shows port disabled/enabled (by the device)
ignore means ignore for alerts (but has to be check by alert rule)
disabled means port is set to disabled in the LibreNMS webui
deleted means the port is still in the DB but pending deletion.

the widget should not show disabled, and deleted devices.  It should show ignore devices, that only applies to alerts.

Right now the bad_if stuff doesn't set disabled, but I plan to extend that to do so.  We could perhaps check is_port_valid() before displaying... might need a little extra to make things better.

File: app/Http/Controllers/Widgets/TopInterfacesController.php (hunk 3 above)
Comment (SourceDoctor) [reply app/Http/Controllers/Widgets/TopInterfacesController.php]: if i change it,  'scopeisActive' is not more camelcase.
if i keep it camelcase 'isActive' cannot be found.

ideas?

Comment (SourceDoctor) [reply app/Http/Controllers/Widgets/TopInterfacesController.php]: changed

**Comment by SourceDoctor** [comment]
@murrant 
scope updated ... hope it matches now better :-)

**Comment by SourceDoctor** [comment]
ok what/where to change/move it to?

**Comment by SourceDoctor** [comment]
@murrant 
could need some ideas what/how the better place would be
or how this would be better implemented.
Or did you do this with your last rewrite already?

**Comment by murrant** [comment]
I implemented part of my thoughts on in the last PR, but the rest requires a re-written port management.

**Pull request closed without merging**